### PR TITLE
General Latitude Longitude grid with precomputed metrics

### PR DESCRIPTION
**Pull request by simone-silvestri**
This PR introduces
 - a general Latitude Longitude Grid with possibility of stretching in all directions (latitude_longitude_grid.jl)
 - possibility of precomputing the LatLon metrics and store them in the grid struct within the constructor
    (precompute_metrics = true/false)
 - a general Rectilinear Grid with an architecture field and possibility of stretching in all directions (rectilinear_grid.jl) 
    This grid will eventually substitute all instances of RegularRectilinearGrid and VerticallyStretchedRectilinearGrid

### CODE REVIEW

File: src/Grids/rectilinear_grid.jl
Comment (tomchor) [reply src/Grids/rectilinear_grid.jl:38]: I think that assumption is fair for the ocean and it probably makes things convenient for the majority of users, but if it complicates the code then I'm very much okay removing it.

**Comment by glwagner** [comment]
@simone-silvestri want me to resolve conflicts?

**Comment by simone-silvestri** [comment]
Let's merge PR #2035 and #2036 first then I can fix this PR

**Comment by simone-silvestri** [comment]
Should be ready to merge.

maybe a note:
- Defining the metrics functions (```Δxᶠᶜᵃ, Δxᶜᶠᵃ, Δxᶠᶠᵃ, Δxᶜᶜᵃ, Δyᶠᶜᵃ, Δyᶜᶠᵃ, Azᶠᶜᵃ, Azᶜᶠᵃ, Azᶠᶠᵃ, Azᶜᶜᵃ````) twice (inside latitude_longitude_grid.jl and spacing_and_areas_and_volumes.jl) is not the most elegant solution so it would be better to define the metrics only in the grid section. This can be done in a next PR where also the Rectilinear grids are unified under one unique RectilinearGrid object.

**Review by tomchor — APPROVED** [review]
Great PR, thanks for the effort.

I'm approving this but I should point out that there are still a few simple comments by @navidcy that haven't been addressed.

Also a quick question: what's left for this to completely replace Regular and VerticallyStretched RectilinearGrids?

**Comment by simone-silvestri** [comment]
Thanks! I'll get the tests to pass completely then merge! I forgot the previous comments, thanks for reminding me, I will address them now. 

For replacing the Regular and Vertically stretched grid it shouldn't be a huge problem but there might be some conflicts. Major ones might be fields like `Δx` being substituted by `Δxᶠᵃᵃ` and `Δxᶜᵃᵃ` or making sure that architecture is always consistent between grid, models and fields

File: benchmark/benchmark_latlon_grid.jl
Comment (glwagner) [inline benchmark/benchmark_latlon_grid.jl:6]:
```suggestion
using Oceananigans.Grids: metrics_precomputed
```

File: src/Grids/latitude_longitude_grid.jl
Comment (glwagner) [inline src/Grids/latitude_longitude_grid.jl:255]: This might be slightly easier to understand --- your call

```suggestion
metrics_precomputed(::LatitudeLongitudeGrid{<:Any, <:Any, <:Any, <:Any, Nothing}) = false 
metrics_precomputed(::LatitudeLongitudeGrid) = true
```

File: src/Grids/latitude_longitude_grid.jl
Comment (glwagner) [inline src/Grids/latitude_longitude_grid.jl:396]: Or perhaps just `allocate_metrics`?

File: src/Grids/latitude_longitude_grid.jl
Comment (glwagner) [reply src/Grids/latitude_longitude_grid.jl:255]: Do we also need a type like

```julia
const PrecomputedMetricsLatLonGrid = LatitudeLongitudeGrid{<:Any, <:Any, <:Any, <:Any, Nothing}
```

?

Comment (simone-silvestri) [reply src/Grids/latitude_longitude_grid.jl:256]: I don't think we need it because that information is not required outside of spacing_areas_and_volumes so maybe a type for that would be a little redundant

File: src/Grids/latitude_longitude_grid.jl
Comment (glwagner) [inline src/Grids/latitude_longitude_grid.jl:101]:
```suggestion
    if precompute_metrics
```

**Comment by tomchor** [comment]
Tests are passing! :tada: 

I don't see any changes in Project.toml though. Don't forget to bump the version before merging.

**Review by navidcy — APPROVED** [review]
That's a big one :) Well done!

**Comment by glwagner** [comment]
# A few benchmarks...

## `benchmark_nonhydrostatic_model.jl`

### `main`

```
                                            Nonhydrostatic model benchmarks
┌───────────────┬─────────────┬─────┬────────────┬────────────┬────────────┬────────────┬──────────┬────────┬─────────┐
│ Architectures │ Float_types │  Ns │        min │     median │       mean │        max │   memory │ allocs │ samples │
├───────────────┼─────────────┼─────┼────────────┼────────────┼────────────┼────────────┼──────────┼────────┼─────────┤
│           CPU │     Float32 │  32 │   7.436 ms │   8.831 ms │   8.815 ms │  11.402 ms │ 2.42 MiB │   7196 │      10 │
│           CPU │     Float32 │  64 │  15.440 ms │  15.923 ms │  16.687 ms │  20.447 ms │ 3.12 MiB │  12808 │      10 │
│           CPU │     Float32 │ 128 │  43.847 ms │  44.717 ms │  44.792 ms │  46.774 ms │ 3.40 MiB │  15383 │      10 │
│           CPU │     Float32 │ 256 │ 280.225 ms │ 289.133 ms │ 289.791 ms │ 307.004 ms │ 3.73 MiB │  35573 │      10 │
│           CPU │     Float64 │  32 │   8.626 ms │   8.873 ms │   8.903 ms │   9.340 ms │ 3.15 MiB │   7192 │      10 │
│           CPU │     Float64 │  64 │  12.859 ms │  13.495 ms │  13.865 ms │  16.695 ms │ 3.95 MiB │  12925 │      10 │
│           CPU │     Float64 │ 128 │  52.882 ms │  54.573 ms │  56.634 ms │  74.680 ms │ 4.31 MiB │  16561 │      10 │
│           CPU │     Float64 │ 256 │ 406.856 ms │ 414.115 ms │ 417.420 ms │ 441.158 ms │ 4.92 MiB │  49543 │      10 │
│           GPU │     Float32 │  32 │   1.932 ms │   2.038 ms │   2.177 ms │   3.359 ms │ 1.42 MiB │   4801 │      10 │
│           GPU │     Float32 │  64 │   1.992 ms │   2.064 ms │   2.167 ms │   3.109 ms │ 1.42 MiB │   4832 │      10 │
│           GPU │     Float32 │ 128 │   2.746 ms │   2.790 ms │   3.045 ms │   5.152 ms │ 1.42 MiB │   5136 │      10 │
│           GPU │     Float32 │ 256 │  15.301 ms │  19.004 ms │  18.634 ms │  19.064 ms │ 1.59 MiB │  15777 │      10 │
│           GPU │     Float64 │  32 │   2.167 ms │   2.272 ms │   2.411 ms │   3.706 ms │ 1.84 MiB │   4775 │      10 │
│           GPU │     Float64 │  64 │   2.260 ms │   2.380 ms │   2.543 ms │   3.984 ms │ 1.84 MiB │   4827 │      10 │
│           GPU │     Float64 │ 128 │   4.640 ms │   4.794 ms │   4.875 ms │   5.688 ms │ 1.86 MiB │   6133 │      10 │
│           GPU │     Float64 │ 256 │  24.612 ms │  36.082 ms │  34.946 ms │  36.210 ms │ 2.19 MiB │  27483 │      10 │
└───────────────┴─────────────┴─────┴────────────┴────────────┴────────────┴────────────┴──────────┴────────┴─────────┘
```

### This PR

```
┌───────────────┬─────────────┬─────┬────────────┬────────────┬────────────┬────────────┬──────────┬────────┬─────────┐
│ Architectures │ Float_types │  Ns │        min │     median │       mean │        max │   memory │ allocs │ samples │
├───────────────┼─────────────┼─────┼────────────┼────────────┼────────────┼────────────┼──────────┼────────┼─────────┤
│           CPU │     Float32 │  32 │   7.850 ms │   8.315 ms │   8.420 ms │   9.770 ms │ 2.42 MiB │   7188 │      10 │
│           CPU │     Float32 │  64 │  14.392 ms │  14.609 ms │  15.268 ms │  20.068 ms │ 3.12 MiB │  12818 │      10 │
│           CPU │     Float32 │ 128 │  43.488 ms │  43.899 ms │  44.401 ms │  49.103 ms │ 3.42 MiB │  16387 │      10 │
│           CPU │     Float32 │ 256 │ 279.247 ms │ 283.817 ms │ 283.421 ms │ 286.571 ms │ 3.63 MiB │  29324 │      10 │
│           CPU │     Float64 │  32 │   8.127 ms │   8.340 ms │   8.484 ms │   9.588 ms │ 3.15 MiB │   7193 │      10 │
│           CPU │     Float64 │  64 │  14.828 ms │  16.011 ms │  15.983 ms │  16.943 ms │ 3.95 MiB │  12805 │      10 │
│           CPU │     Float64 │ 128 │  53.118 ms │  54.351 ms │  54.726 ms │  60.187 ms │ 4.33 MiB │  17844 │      10 │
│           CPU │     Float64 │ 256 │ 406.215 ms │ 412.665 ms │ 414.436 ms │ 427.027 ms │ 5.05 MiB │  51705 │      10 │
│           GPU │     Float32 │  32 │   3.327 ms │   3.428 ms │   3.627 ms │   5.105 ms │ 1.44 MiB │   5292 │      10 │
│           GPU │     Float32 │  64 │   3.115 ms │   3.227 ms │   3.363 ms │   4.620 ms │ 1.44 MiB │   5338 │      10 │
│           GPU │     Float32 │ 128 │   4.711 ms │   4.874 ms │   4.937 ms │   5.755 ms │ 1.45 MiB │   5750 │      10 │
│           GPU │     Float32 │ 256 │  22.139 ms │  23.264 ms │  23.625 ms │  27.468 ms │ 1.46 MiB │   6749 │      10 │
│           GPU │     Float64 │  32 │   3.348 ms │   3.436 ms │   3.573 ms │   4.861 ms │ 1.86 MiB │   5280 │      10 │
│           GPU │     Float64 │  64 │   3.391 ms │   3.428 ms │   3.585 ms │   4.839 ms │ 1.86 MiB │   5330 │      10 │
│           GPU │     Float64 │ 128 │   6.782 ms │   6.942 ms │   7.072 ms │   8.233 ms │ 1.87 MiB │   6100 │      10 │
│           GPU │     Float64 │ 256 │  40.043 ms │  40.948 ms │  41.482 ms │  44.300 ms │ 1.91 MiB │   8566 │      10 │
└───────────────┴─────────────┴─────┴────────────┴────────────┴────────────┴────────────┴──────────┴────────┴─────────┘
```

It looks like large GPU models may be affected. @simone-silvestri what do you think?

**Comment by simone-silvestri** [comment]
It is the effect of `wait(device(arch), events)` vs the `wait(events)` in `fill_halo_regions.jl` line 64

(On main just substituting that line)

```
julia> benchmark_nonhydrostatic_model(GPU, Float64, 128)
BenchmarkTools.Trial: 10 samples with 1 evaluation.
 Range (min … max):  4.784 ms … 22.835 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     4.836 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   6.634 ms ±  5.692 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                           
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄ ▁
  4.78 ms        Histogram: frequency by time        22.8 ms <

 Memory estimate: 1.86 MiB, allocs estimate: 6061.
```


vs 

```
julia> benchmark_nonhydrostatic_model(GPU, Float64, 128)
BenchmarkTools.Trial: 10 samples with 1 evaluation.
 Range (min … max):  6.907 ms … 25.036 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     7.512 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   9.426 ms ±  5.532 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

   █▄                                                         
  ▆██▁▁▁▁▁▆▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▆ ▁
  6.91 ms        Histogram: frequency by time          25 ms <

 Memory estimate: 1.86 MiB, allocs estimate: 5563.
```

I ll investigate more this wait function today.

**Comment by simone-silvestri** [comment]
It is the effect of `wait(device(arch), events)` vs the `wait(events)` in `fill_halo_regions.jl` line 64

(On main just substituting that line)

```
julia> benchmark_nonhydrostatic_model(GPU, Float64, 128)
BenchmarkTools.Trial: 10 samples with 1 evaluation.
 Range (min … max):  4.784 ms … 22.835 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     4.836 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   6.634 ms ±  5.692 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                           
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄ ▁
  4.78 ms        Histogram: frequency by time        22.8 ms <

 Memory estimate: 1.86 MiB, allocs estimate: 6061.
```


vs 

```
julia> benchmark_nonhydrostatic_model(GPU, Float64, 128)
BenchmarkTools.Trial: 10 samples with 1 evaluation.
 Range (min … max):  6.907 ms … 25.036 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     7.512 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   9.426 ms ±  5.532 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

   █▄                                                         
  ▆██▁▁▁▁▁▆▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▆ ▁
  6.91 ms        Histogram: frequency by time          25 ms <

 Memory estimate: 1.86 MiB, allocs estimate: 5563.
```

I ll investigate more this wait function today.

**Comment by glwagner** [comment]
Is it possible that the problem lies elsewhere, so that changing `wait` just masks the real issue rather than solving it?

**Comment by simone-silvestri** [comment]
For sure, that is always an option

**Pull request merged**
